arch/wasm/bits/float.h

```diff
@@ -0,0 +1,20 @@
+#ifdef __FLT_EVAL_METHOD__
+#define FLT_EVAL_METHOD __FLT_EVAL_METHOD__
+#else
+#define FLT_EVAL_METHOD 0
+#endif
+
+#define LDBL_TRUE_MIN 6.47517511943802511092443895822764655e-4966L
+#define LDBL_MIN 3.36210314311209350626267781732175260e-4932L
+#define LDBL_MAX 1.18973149535723176508575932662800702e+4932L
+#define LDBL_EPSILON 1.92592994438723585305597794258492732e-34L
+
+#define LDBL_MANT_DIG 113
```

Comment:
@sunfishcode, do you know how Wasm has 80-bit long-double support? I would have expected long-double to be the same as double, and was surprised to find that it's not. Does clang emulate the higher-precision, or is it fictitious?

I got these constants here by dumping Clang's predefined macros (as requested by Rich Felker in his review on the Musl mailing list).

Reply:
Clang is currently configured to make long-double be a 128-bit IEEE-754 quad-precision type. It is supported through software emulation in compiler-rt (it calls __addtf3 for addition, and so on).

The decision to do this is debatable. On one hand, making long double the same as double makes long double effectively useless, and making it longer makes it at least useful for some purposes (even if they are fairly narrow). On the other, musl's printf promotes all floating-point values to `long double` to avoid duplicating code, and this means that all floating-point formating goes through slow software-emulated paths, and it increases code size for any program that calls printf.

I expect we'll eventually have custom versions of printf that will disable disable various bits of floating-point support, because many applications won't need everything (and the nature of printf format strings makes it difficult to dce unused formatting code), and if we do, this would significantly mitigate the problems. However, the utility of 128-bit long double is admittedly obscure, so it's not completely certain that the benefits outweigh the costs.

Reply:
When you say "Clang is currently configured", I assume that's specific to the WebAssembly backend? Given that GCC on x86 doesn't give you 128-bit quad-precision floats, the need for that level of precision seems remote.

I don't know - maybe people really would welcome having that high precision available in Wasm via Clang. But, to me it just seems like a pitfall.

Something to think about, thanks for explaining. I'd vote to have double match long-double on Wasm, at least until the demand for it is strong enough. Since we don't yet have dynamic linkage, changing the ABI isn't too big a deal.

Reply:
I've found https://github.com/kripken/emscripten/issues/4340 now. I see Derek was thinking of going back to 64-bit long-double, but that doesn't seem to have happened. Oh well - up to you guys!